metricbeat/module/mssql/db/db.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/pkg/errors"
+	"sync"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("mssql", "db", New, mb.DefaultMetricSet())
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	*mssql.MetricSet
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The mssql io metricset is experimental.")
+
+	config := struct{}{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	//TODO Db must be gracefully closed
+
+	metricSet, err := mssql.NewMetricSet(base)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating mssql metricset")
+	}
+
+	ms := &MetricSet{MetricSet: metricSet}
+
+	return ms, nil
+}
+
+func newConcurrentQueries(db *sql.DB, qs []string) *concurrentResults {
+	qr := &concurrentResults{
+		queries: qs,
+		db:      db,
+	}
+
+	qr.Add(len(qs))
+
+	return qr
+}
+
+type concurrentResults struct {
+	queries []string
+	maprs   []common.MapStr
+	error
+	sync.Mutex
+	sync.WaitGroup
+	db *sql.DB
+}
+
+func (c *concurrentResults) doQueries() {
+	for _, q := range c.queries {
+		go c.doEventsFetcherQuery(fmt.Sprintf("SELECT * FROM sys.%s;", q))
+	}
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch() ([]common.MapStr, error) {
+	result := newConcurrentQueries(m.Db, []string{
+		"dm_db_file_space_usage",
+		"dm_db_fts_index_physical_stats", //TODO Returns nothing with empty db
+		//TODO dm_db_log_info
+		"dm_db_log_space_usage",
+		//TODO dm_db_log_stats
+		"dm_db_partition_stats",
+		"dm_db_persisted_sku_features", //TODO Returns nothing with empty db
+		"dm_db_session_space_usage",
+		"dm_db_task_space_usage",
+		"dm_db_uncontained_entities", //TODO Returns nothing using the driver. Works with sqlcmd
+	})
+	result.doQueries()
+	result.Wait()
+
+	return result.maprs, result.error
+}
+
+func (c *concurrentResults) doEventsFetcherQuery(q string) {
+	defer c.Done()
+
+	maprSlice, err := doQuery(c.db, q)
+	if err != nil {
+		c.Lock()
+		defer c.Unlock()
+		if c.error == nil {
+			c.error = err
+		} else {
+			c.error = errors.Wrap(c.error, err.Error())
```

Comment:
I think you are looking for `multierror` here, not `errors.Wrap` :slightly_smiling_face: